### PR TITLE
Fix crash on datafields

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Related Issue

Fixes the crash reported on https://github.com/PyCQA/pylint/issues/4963
Tests added there.

### CODE REVIEW

File: astroid/brain/brain_dataclasses.py [hunk 1]
```diff
@@ -231,19 +232,20 @@ def infer_dataclass_attribute(
...
-    default_type, default = _get_field_default(field_call)
-    if not default_type:
-        yield Uninferable
-    elif default_type == "default":
-        yield from default.infer(context=ctx)
-    else:
-        new_call = parse(default.as_string()).body[0].value
-        new_call.parent = field_call.parent
-        yield from new_call.infer(context=ctx)
+    if isinstance(node.parent, (AnnAssign, Assign)):
```
Comment (Pierre-Sassoulas) [inline astroid/brain/brain_dataclasses.py:238]:
```suggestion
    if not isinstance(node.parent, (AnnAssign, Assign)):
        return
```
Unless it does not works with yield ? It permits to simplify the number of things to keep in mind when reading the code and reduce the diff and future conflict.

**Review by Pierre-Sassoulas — COMMENTED** [review]
👍

**Comment by DanielNoord** [comment]
Also changed the typing of the return!

File: astroid/brain/brain_dataclasses.py [hunk 2]
```diff
@@ -231,9 +232,11 @@ def infer_dataclass_attribute(
 
 
 def infer_dataclass_field_call(
-    node: AssignName, ctx: context.InferenceContext = None
-) -> Generator:
+    node: Call, ctx: context.InferenceContext = None
```
Comment (DanielNoord) [inline astroid/brain/brain_dataclasses.py:235]:
```suggestion
    node: Call, ctx: Optional[context.InferenceContext] = None
```

File: astroid/brain/brain_dataclasses.py [hunk 3]
```diff
@@ -231,9 +232,11 @@ def infer_dataclass_attribute(
 
 
 def infer_dataclass_field_call(
-    node: AssignName, ctx: context.InferenceContext = None
-) -> Generator:
+    node: Call, ctx: Optional[context.InferenceContext] = None
+) -> Optional[Generator]:
     """Inference tip for dataclass field calls."""
+    if not isinstance(node.parent, (AnnAssign, Assign)):
+        return
```
Comment (cdce8p) [inline astroid/brain/brain_dataclasses.py:239]: I think this should be `raise UseInferenceDefault` instead. Please test it before changing it as I'm not 100% certain.

File: astroid/brain/brain_dataclasses.py [hunk 4]
```diff
@@ -231,9 +232,11 @@ def infer_dataclass_attribute(
 
 
 def infer_dataclass_field_call(
-    node: AssignName, ctx: context.InferenceContext = None
-) -> Generator:
+    node: Call, ctx: Optional[context.InferenceContext] = None
+) -> Optional[Generator]:
```
Comment (cdce8p) [inline astroid/brain/brain_dataclasses.py:236]:
```suggestion
) -> Generator:
```
The return type will then only be `Generator`.

**Review by cdce8p — CHANGES_REQUESTED** [review]
I saw that you opened a PR in pylint with regression tests. It would be great if you could add a test for astroid as well.

**Comment by DanielNoord** [comment]
Not quite sure what the tests should be. I don't really know how `astroid` works, but saw the issue in `pylint` and saw that it was quite easily fixable.

Please let me know what the tests should be and then I will see if `raise UseInferenceDefault` also works.

File: tests/unittest_brain_dataclasses.py [hunk 5]
```diff
@@ -643,3 +643,39 @@ class A:
...
+    from typing import cast
+
+    @dc.dataclass
+    class A:
+        attribute: int = cast(int, field(default_factory=dict))
+
+    """
+    )
+    inferred = node.inferred()
+    assert len(inferred) == 1
```
Comment (cdce8p) [inline tests/unittest_brain_dataclasses.py:663]:
```suggestion
    assert len(inferred) == 1 and isinstance(inferred[0], nodes.ClassDef)
    assert "attribute" in inferred[0].instance_attrs
```
Just testing that `attribute` is indeed an `instance_attr` should be enough.

File: tests/unittest_brain_dataclasses.py [hunk 6]
```diff
@@ -643,3 +643,39 @@ class A:
...
+def test_enclosed_field_call(module: str):
+    """Test inference of dataclass attribute with a non-annotated field call in another function call"""
+    node = astroid.extract_node(
+        f"""
+    from {module} import dataclass, field
+    from typing import cast
+
+    @dc.dataclass
+    class A:
+        attribute = cast(int, field(default_factory=dict))
```
Comment (cdce8p) [inline tests/unittest_brain_dataclasses.py:676]: Why not combine both tests by adding this as `attribute2` in the first test. Not really necessary to have two separate once for it.

**Review by cdce8p — COMMENTED** [review]
Had a bit of time to test it. `raise UseInferenceDefault` will work.

File: tests/unittest_brain_dataclasses.py (hunk 6 above)
Comment (DanielNoord) [reply tests/unittest_brain_dataclasses.py:676]: Good idea! Going to do this now!

**Comment by DanielNoord** [comment]
Turns out I did not know enough about `dataclasses`. The second test was non-sensical and did not cover anything not already covered by the other tests so I removed it.

**Review by cdce8p — APPROVED** [review]

**Pull request merged**